### DISCUSSION TITLE
NES-Connectors Vision (Post Milestone 2)

### DISCUSSION THREAD

**Discussion by alepping** [nebulastream/nebulastream#10] [Ideas]
# Description
This discussion is about how we could work towards a NebulaStream version that can be used without the need for any config. 

## Goal
Develop a concrete plan for how to change NebulaStream so that we can start it with zero configuration and that it provides a similar feeling to using a batch database like SQLite

## Problem
**Implementation:**
Currently, it is quite cumbersome to set up a NebulaStream instance that one can query and that one can the the query results of.
- we need to setup a coordinator with a yaml file that contains a logical source
- we need to setup a worker that correctly connects to that coordinator with a yaml configuration that defines a physical source
- we need to start the coordinator
- we need to start the worker
- we need to write a query in C++
- we need to define a sink descriptor within the query that is later converted to a sink in nebulastream

Many of these steps require different interfaces.
 
**Usability:**
In a batch database like SQLite, a user needs to do the following to query a database from scratch:
- start the system `sqlite`
- create a database (DDL SQL syntax)
- create a table in the database (DDL SQL syntax)
- add data to the table (DML SQL syntax)
- query data from the table (DQL SQL syntax)

While these are many steps, they are all based on SQL syntax. This enables us to create a `do_everything.sql` file that contains all the SQL instructions from a `.sql` file and sets up a fully working database that we can interact with.
We could potentially have a similar setup in NebulaStream:
- start a worker and a coordinator
- create a logical schema (DDL - creating a table) (might be used by many physical sources (e.g., heart monitors in many ICU beds))
- create a physical source (DML - inserting data) (one specific heart monitor in one specific ICU bed)
( we can skip creating a sink if the default sink is stdout)
- submit a query (DQL - querying data)

The observation is that our process can be very similar and almost as simple as the process required to query a batch database (simple use case, not regarding massive distribution, etc.).
Currently, I believe that if all the above steps would be possible using a single language (e.g., SQL), using NES would feel natural for database people.
The [SQL API used by Flink](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/create/) demonstrates how this could look

## Vision
It would be ideal if we could start NebulaStream without configuration. The default configs enable workers and coordinators to find each other. 
- to setup a logical source, we can issue a query (DDL). 
- to setup a physical source, we can issue a query (potentially the coordinator broadcasts the potential physical source to all workers and finds out which workers could support the physical source. Not optimal concerning performance, but user friendly (~DML)
- to setup a (physical) sink, we can issue a query (~DML)
- to submit a query, we can issue a query (DQL)

In principle, we can write all these queries in one file and just submit it to NebulaStream (similar to a .SQL file).
Additionally, a default (logical&physical) source would also make setting up and using NebulaStream for the first time very convenient.

###
Queries could look similar to the following:
```sql
-- Define a logical source that is a container for physical sources.
-- The logical source defines the schema that the physical sources must follow.
CREATE LOGICAL SOURCE nesCityHospital (
    PatientID UINT32,
    LastName Text,
    FirstName Text,
)

-- Define logical sink that is a container for physical sinks.
-- We don't require a schema for sinks, since we can/should infer it.
CREATE LOGICAL SINK grafana_board;

-- Attach physical MQTT source to logical source ('nesCityHospital' Table)
-- The physical MQTT source INSERTs data into the nesCityHospital schema.
ATTACH PHYSICAL SOURCE nesCityHospital (
    connector=mqtt,
    format=xml,
    url=ws://172.31.0.11:9001,
    topic=solarPanels,
    qos=2
)

-- Attach a physical sink to the logical sink 'grafana_board'
ATTACH PHYSICAL SINK grafana_board (
    connector=mqtt,
    format=json,
    url=ws://172.31.0.12:9003,
    topic=grafana_data,
    qos=1
)
```

**Comment by klauck** [comment]
I agree.
Something like a NebulaStream-Console would be nice for trying out our system.
Once we can run standalone workers, this should not be too hard.
The question is whether we want to support a SQL-like language right from the start or use an easier to parse stream/table definitions and queries first.

**Reply by zeuchste** [reply]
Hi maybe you could start in the DataSource.cpp there is a lot to investigate and this is the running class for all sources

**Comment by zeuchste** [comment]
Hi Aljoscha the idea is very cool and the vision is really what we need. The question is then what we need to change to make it happen. Some points:

- logical source can be atm already created over REST, with physical sources I am not sure. The main problem is here that this breaks our assumption that user do not know anything about the node and cannot interact with them directly. So there is no command like: "Say node X to do Y" as from the outside world this is not exposed (we hide it behind the logical stream for example). The question if if we are willing to relax on this property. This is also a security problem.
- The broadcast idea I wrote also in the other document as a possible solution but the question remains how can a worker decide if he 1) can do this (ok by checking file path for example but how about TPC?) and also 2) if he wants to give this data out? How can he gives his permission then?
- The idea of a DML is nice but how would this look like? Do we really think ppl will learn our own DDL? Maybe we can slightly modify existing once to lower the entrance barrier
-

**Reply by alepping** [reply]
Hey Steffen,
Thanks for the reply and the interesting concerns!
I also thought about your **first point** and I thought about relaxing the assumption. However, would it actually be relaxing? You said:
> "... this breaks our assumption that user do not know anything about the node and cannot interact with them directly."

This is not true though, depending on who you consider the user. Currently, a 'user' has to configure each individual worker (node) and tell them which physical sources they support. This is far from the user not knowing about nodes. If you differentiate between backend users that setup the infrastructure and frontend users that just submit analytical queries, then yes, the frontend user does not need to know about physical sources. But, what if the frontend user wants a new source? Do they always need to ask the backend user first? What if a physical source disappears? Should this not be transparent to the frontend user?
A possible solution would be that frontend users know about nodes and can address them directly (again, backend users know about nodes anyway). Also, we expose the topology to frontend users too currently.

Concerning the **second point**, I think a node can 'easily' detect whether it can support a physical source. I think the bigger problem here is if a physical source could exist on multiple nodes. Say a CSV file exists on the same path for several nodes. It seems inevitable that the (frontend) user needs specify the node(s). The security concern would be authorization then, right? I think this is a bigger problem in general that we need to think about.

I think people don't need to learn our DSL, but we should use a language that they are used to and at most extend it.

**Reply by zeuchste** [reply]
For the first point, I would not distinguish between front end and backend user for me there is the "user"/owner of a node that currently specify how and what he wants to contribute at startup of the node. If we bind over the coordinator this means every user can force a worker to start a physical source. This is a security disaster from two points. First every user can start over rest a TCP source on any node in the system and then send them over the TCP port an infected package. Second, the owner of the node does not decide anymore what he wants to contribute as again users can force this ofer the Rest. As we have no user/right management, it would be possible for all users to do this atm.

For the second point, I do not know, I mean every node can have a physical source e.g. a TCP source open on a port. What could be good for testing would be to send the csv file to the coordinator and then this one distributes it over the nodes and start the sources for that. But this is just for testing as in the normal setup mode, the node will have a real sensor attached and this will give the data. I think we cannot come around the fact that our system is source driven and that only the node where the sensor/source is attached really know what it can and want to contribute.

If you want to continue using the backend/frontend user then lets define them first what are their duties and rights

**Reply by alepping** [reply]
I agree with the security problem. However, this is a different topic from my point of view. We need authentication and authorization in our system. This means, that certain users cannot deploy queries on certain nodes or read data produced by certain nodes. This also means that a user may only bind a physical source to a specific worker if they have the correct rights (are authorized to do so).
In my opinion, we should handle this problem on the user rights level (also part of worker configuration) and not on the physical source configuration level.
As I said above, it is very important point, but it should be a separate discussion from my point of view.

**Reply by zeuchste** [reply]
yes but if you want to unlock this feature we then need user rights and user management which we do not have atm :)

**Reply by keyseven123** [reply]
Do we really need full user rights and user management? 
I agree that we should come up with a user management but for the first and especially the second milestone, we can assume that everyone has every right and permission. 
Maybe w can implement something together with the NEEDMI people/project in a third milestone.

**Reply by zeuchste** [reply]
Hi Nils, yes you are right we maybe do mix up things here a lot, sometimes we talk about the next milestone and sometimes about the open source release and sometimes about the longer vision.

In this case, lets start with your vision of user management that everybody can do everything, but I think it would be good to still keep in mind that as a consequence of the external binding of physical sources we need than a simple form of user management or other means for the open source release as otherwise we make the entire system a big security risk for all users

**Comment by alepping** [comment]
@zeuchste said in discussion #8:

Hi in general this sound good but there are some important points missing to understand it fully:

1. When you say source registry where is this on a node or the coordinator? 
2. The problem with sources is that they are totally node specific. Imagine a TPC source or a CSV source the logical source is easy to maintain globally but the physical source is different for each node, i.e., different ip and different file path
3. It is unclear when exactly the sources are registered and where. So at startup over the coordinator via rest or on node startup on the command line or on node setup via config file
4. One major problem that we always had with sources is that users where overwhelmed with the parameters they have to specify, here we should pick very good defaults (like every second), document them, and only let the user specify the bare minimum
5. A different idea would be to turn around the source registration. So instead of the worker starts and says this are my sources to the coordinator, the coordinator sends a list of sources and the worker decides which one he can offer (fix idea not well thought) 
6. In general we need so separate here the use of the source (cumbersome over config files) and the addition of new sources (cumbersome as we have to attach many files). Both should be considered

**Reply by alepping** [reply]
@ls-1801 replied to @zeuchste in discussion #8:

I like @zeuchste idea for the coordinator reaches out to all worker about which physical sources it supports.

But for some sources (actually almost all) the source type is not enough and whether a specific source can be instantiated on a worker depends on the configuration. E.g a filesource may require a specific file to exist.

However I think it would also make sense to ask for support, checking if plugin is loaded, and creating an Error when trying to start a query with specific configuration.

For our first milestone I would keep it simple and just accept arbitrary source configuration and report an error if the source cannot be established, I think the plugin system and solid data source implementation is more important

**Reply by alepping** [reply]
@ls-1801 I completely agree, the type is not enough. When the coordinator reaches out to all nodes, the coordinator would need specify all necessary parameters to instantiate the physical source. Each node could then check if, given the parameters, they can support the physical source and reply.

**Comment by alepping** [comment]
@PhilippGrulich said in discussion #8:

While reading I had the following idea. Maybe it would be easier to manage both logical and physical sources at the coordinator.
As a result:
- all queries are bound to specific physical sources when they arrive at the worker.
- the worker is stateless and keeps no state of physical sources. 
- the coordinator is the only entity, which manages the mapping from logical to physical sources.
- we could have two ways to register a new physical source:
1. via the worker configuration -> basicall the worker just tells the coordinator which sources it has.
2. via the coordinator -> for example using some DML statement: 
**BIND FILE("PATH") AT "NODE2" TO STREAM "X"** 

This would also have the benefit that we could support "flexible" physical sources. 
For example, some sources like KAFKA could be placed on any node.

**Reply by alepping** [reply]
@zeuchste replied to @PhilippGrulich in discussion #8:

@PhilippGrulich yes your approach is also valid and good. But again, currently we do have the assumption that outside world do not know anything on inside world. If you let the binding happen over the coordinator, this is violated (which is maybe not bad but we need then to accept this as a new property).

The stateless idea of physical config is maybe not possible as the worker has to start and stop and interact with the source locally, how will he do it if he does not know about it.

After reading the entire discussion, I think the main thing we need to change is that we allow DML like BIND over the coordinator for physical sources which is currently not possible. This would enable us to for example use a setup script like @alepping suggested.

@ls-1801 thanks for supporting the idea of a coordinator ask the sources what they can contribute. This would be really cool but again how is this organized how can a worker decide that yes he can open a TCP port 5555 but if he wants to. Also the files can lie on different path maybe. If anybody has a cool idea to solve this, I think this would be the best solution.

**Reply by alepping** [reply]
Hey @PhilippGrulich , thanks for sharing your ideas.
I guess I came to the same conclusion that you reached.
Additionally I see two more ways to register physical sources. 
1. The broadcast solution discussed above (coordinator asks all nodes whether they can support a specific physical source).
2. Physical sources can register themselves at nodes. In particular, each node has an endpoint that can be used to register a specific physical source

**Reply by PhilippGrulich** [reply]
Sounds good. My only concern is the coordination between worker and coordinator. 
If the worker keeps state about the physical sources this introduces a lot of pain points as the coordinator has to keep the state in sync. The question is if this is actually required and why?
For example, why would a worker need to know the physical sources at runtime?

**Reply by alepping** [reply]
Hey @PhilippGrulich ,
if I understand you correctly, you are saying that is sufficient to supply the worker with the information of a physical source with the query. That is, a worker doesn't know anything about physical sources. Only when the coordinator supplies the worker with a query, the worker instantiates a physical source (to read data from somewhere) and with a sink (to write data somewhere).
Is that correct?

**Reply by zeuchste** [reply]
to be honest I do not really know where this leads here now maybe @alepping you can summarise all the changes and protocols here in a condensed way?

**Reply by PhilippGrulich** [reply]
> Hey @PhilippGrulich ,
> 
> if I understand you correctly,...
> 
> Is that correct?

Yes exactly. In this case the worker is quite stupid and the coordinator has all the control.

**Reply by alepping** [reply]
@zeuchste and @PhilippGrulich, check the 'solution' section of discussion #8 for a description of where this leads.

**Comment by zeuchste** [comment]
### Internal vs External
One thing to consider is that all the things we discuss here are only relevant because we want to test our system. We have internal sources like Csc, file, etc which only purpose is for us to test the system. No end user would ever want to stream a csv in a real world setup as there the data comes from real sensors attached to the nodes or via Kafka etc. 

Thus, I think we can maybe change the internal sources as much as we want as they are only for testing purposes. Maybe we should consider this in our discussion too.

**Reply by keyseven123** [reply]
Hi,
If necessary we can change the internal test sources, but in my opinion, this should be our last resort. We should also try to get away with mocking as much as possible without having to actually implement a `CSVTestSource` or something like that. 
This point should be taken into consideration of how we structure our tests

**Reply by zeuchste** [reply]
Hi this is not only on tests in the sense of testing the code. It is also how other ppl come to NES and test it first before they really do it in production (where they would have external sources). So it is more about first time usage or lower the barrier of getting started fast.

**Comment by zeuchste** [comment]
### New Idea: External Source only

One idea that comes to my mind to really make this source thing more clear is that we remove all internal sources like CSV file etc and only allow external sources. The definition of an external source is that is pushed data to the system, like in a TPC, ZMQ, source as this this how the real system will work. For Testing and easy usage we add a new component to NES lets call it Source Driver that is capable of for example getting a CSV or file and push it with certain conditions to the waiting external source. 

This would clearly separate the weird source configuration and what the system really does. It could then also be integrated into the test harness. Another benefit is that we can get rid of the ugly DataSource.cpp that wakes up at different points in time and actively pull a buffer from a source. This makes no sense in real world systems and is may to complicated. The only thing that a source then needs to do is if a package comes, create a buffer and call the buffer manager. No spinning thread is required.

What do you think @alepping @keyseven123 ?

**Reply by zeuchste** [reply]
This would replay simplify the entire sources efficiently.

If a query is submitted, it opens just an external source on each node and waits for packages.

Then if the user then wants to test/simulate a workload he can for example start the source driver:

`./sourceDriver --input=test.csv --frequenceInTupsPerSecond=100 --output=localhost:8081/source3`

This would solve the majority of problems that we have with the sources in my opinion and makes a clear separation between sources in NES and the data ingestion

**Reply by adrianmichalke** [reply]
Hi, I'm not sure if I understand this idea correctly. Maybe explain in more detail were you see concrete advantages in this approach.

Having an additional executable that you need to deploy would be a big disadvantage for me. Also, CSV support would be a main feature in real systems for me.
I'm also wondering about batch support. Would every table-like source be an internal source and thus not core part of our sources?

**Comment by zeuchste** [comment]
### Intermediate Agreement
After discussing the topics and points in this discussion  {with Adrian and Aljoscha) , we concluded on the following points

- [ ] The current source is hard to setup because it requires a config file with weird parameters on every device:
-- Solution: We specify one DML statement over the coordinator that creates the logical source AND specifies the parameter for every participating physical source (like Kafka on port X). The user then only needs to start the node engine and specifies the logical source it wants to contribute and gets then the config from the coordinator and initiates it. Alternatively, User can at runtime specify that a particular node now contributes to that stream by another DML statement.

- [ ] The current implantation using internal and external sources are rather artificial
-- Internal sources are only for testing as a real-world system would only use external source like Kafka that pushed data to the system. By removing the internal sources form the NES core and transfer them to an own sourceDriver, we externalise the logic for that such that users understand that if they have a real steam they can use connect it to NES via an external source or if they want to test stuff, they use the driver to simulate the stream first

**Reply by zeuchste** [reply]
One solution would be to build the driver on top of https://github.com/streamthoughts/kafka-connect-file-pulse that gives us everything we need then we can use the Kafka source for NES and that's it

**Comment by alepping** [comment]
@adrianmichalke michalke said in #64:
> Cool example [referring to the query example for attaching physical sources/sinks], I like it a lot!
> Interesting, that we define a logical sink. Is it like a 'multiplexer' to send to multiple ph. sinks?

@alepping replied:


@adrianmichalke
> Thank you!
> Concerning the logical sink, I had two ideas in mind:
> - physical sink as a variable: assigning a name to a specific physical sink, with a specific format, with a specific configuration. As a result, we don't have to define (DDL) physical sinks in analytical (DQL) queries.
> - producing one result tuple buffer with many different sinks

@zeuchste replied:
> Interesting concept of a logical sink, what would be required is that all data follows the same schema. But if we infer the schema and two queries want to push in the same sink with different schema I think there could be problems. Also I try to think of real cases where two queries would share the same sink and in addition if they would, would the query sharing not identify this and do this automatically?

> What is not clear to me in the example is how we now say node x gets physical source y. How would the DDL for his look like

**Reply by alepping** [reply]
@zeuchste ,
> What is not clear to me in the example is how we now say node x gets physical source y. How would the DDL for his look like

I realized this too. I updated the example:
> ATTACH PHYSICAL SOURCE nesCityHospital (
>    nodeid=1,

When attaching a physical source, a nodeid must be specified.

**Reply by alepping** [reply]
> I try to think of real cases where two queries would share the same sink and in addition if they would, would the query sharing not identify this and do this automatically?

The main idea is syntactic sugar for users. Say you have a query that sends temperature data to four different sinks (local file, mqtt grafana, cloud via kafka, TCP to another node) then you can group these sinks and simply use the logical sink name in the query.
This is also already true for a single query. Currently, we have to redefine the entire physical sink, everytime we use it, even it is the exact same physical sink. This is true for the current NEEDMI queries.


> But if we infer the schema and two queries want to push in the same sink with different schema I think there could be problems

The way I envisioned it above was as a syntactic sugar mechanism to group physical sources for use in a single query. If you start two different queries using the same 'logical sink', the coordinator would infer two different schemas, one for each query, and everything should work out.
As far as I can see this is already true for a single query. There might be two different sinks in a single query that receive two different schemas. However, these different sinks must be part of the query. Thus, the coordinator can infer the different schemas for both sinks.

**Comment by fwc** [comment]
## Dynamic Definition of Sources/Sinks

I do not see, if/where the proposed syntax describes on which node the source/sink should be located.


```
-- Attach a physical sink to the logical sink 'grafana_board'
ATTACH PHYSICAL SINK grafana_board (
    connector=mqtt,
    format=json,
    url=ws://172.31.0.12:9003,
    topic=grafana_data,
    qos=1
)
```
I think it would be nice to say something like: "Pls read from this sensor device / named pipe / ... from all those nodes".

How about something like the following, were defining a source is similar to inserting into a virtual table, the topology can be queried and we can even query special information about workers via a function-like thingy:

```
insert into physical_sources (worker_id, source)
select 
    worker_id, '<json or whatever for source description>'
from topology where region = 'Berlin' and localFileExists('/dev/my_funny_sensor_device');
```
